internal/rpcnetworkclient/client.go

```diff
@@ -0,0 +1,219 @@
+func (e *NetworkError) IsBadSequence() bool {
+	result, err := e.ResultXDR()
+	if err != nil {
+		return false
+	}
+	return result.Result.Code == xdr.TransactionResultCodeTxBadSeq
+}
+
+// ResultXDR parses and returns the transaction result XDR.
+func (e *NetworkError) ResultXDR() (xdr.TransactionResult, error) {
+	var result xdr.TransactionResult
+	resultXDR, err := e.ResultString()
+	if err != nil {
+		return result, err
+	}
+	if err := xdr.SafeUnmarshalBase64(resultXDR, &result); err != nil {
+		return result, err
+	}
+	return result, nil
+}
+
+// IsTimeout returns true if the error indicates a timeout occurred.
+func (e *NetworkError) IsTimeout() bool {
+	return e.timeout
+}
+
+// ResultString returns the result string from the error, if available.
+func (e *NetworkError) ResultString() (string, error) {
+	if e.resultXDR == "" {
+		return "", fmt.Errorf("no result XDR available")
+	}
+	return e.resultXDR, nil
+}
+
+// DiagnosticEventStrings returns the diagnostic event XDR strings, if available.
+func (e *NetworkError) DiagnosticEventStrings() []string {
+	return e.diagnosticEventsXDR
+}
+
+// Error implements the error interface.
+func (e *NetworkError) Error() string {
+	return e.err.Error()
+}
+
+// Unwrap returns the underlying error.
+func (e *NetworkError) Unwrap() error {
+	return e.err
+}
+
+// NetworkClient wraps an RPC client and implements the internal.NetworkClient interface.
+type NetworkClient struct {
+	client *rpcclient.Client
+}
+
+// NewNetworkClient creates a new NetworkClient wrapping the given RPC client.
+func NewNetworkClient(url string, httpClient *http.Client) *NetworkClient {
+	client := rpcclient.NewClient(url, httpClient)
+	return &NetworkClient{client: client}
+}
+
+// SubmitTransaction submits a transaction using the underlying RPC client.
+// It blocks until the transaction is finalized (SUCCESS or FAILED) or times out after 30 seconds.
+func (r *NetworkClient) SubmitTransaction(txXDR string) error {
+	// TODO: Pass context down from the request upstream. See https://github.com/stellar/friendbot/issues/22
+	ctx, cancel := context.WithTimeout(context.Background(), submitTransactionTimeout)
+	defer cancel()
+
+	request := protocol.SendTransactionRequest{
+		Transaction: txXDR,
+	}
+
+	response, err := r.client.SendTransaction(ctx, request)
+	if err != nil {
+		return &NetworkError{err: err}
+	}
+
+	// If the transaction was rejected immediately, return the error
+	if response.Status == statusError {
+		return &NetworkError{
+			err:                 fmt.Errorf("transaction rejected"),
+			resultXDR:           response.ErrorResultXDR,
+			diagnosticEventsXDR: response.DiagnosticEventsXDR,
+		}
+	}
+
+	return r.pollTransactionStatus(ctx, response.Hash)
+}
+
+// pollTransactionStatus polls GetTransaction until the transaction is finalized
+// (SUCCESS or FAILED) or the context times out.
+func (r *NetworkClient) pollTransactionStatus(ctx context.Context, txHash string) error {
```

Comment:
We should add this to the client

Reply:
Something to consider, but I think the way I'm doing the polling here is probably not best practice. Doing this polling is fine in Friendbot when we don't really care if the transaction fails or succeeds, not in the same way that a production application cares. Other applications at scale are more likely to process payments in a queue, match them up to previously submitted ones, not in memory. Instead, using some sort of persistent database.

I think if we put this into the RPC client, we'd be encouraging the same type of practices that Horizon's Synchronous API encouraged, where folks don't really handle the edge cases for when a transaction fails or succeeds outside of a short window.

Reply:
Yeah I don't necessarily mean this exact implementation, but the whole "submit and poll" pattern is so common that we shouldn't make every client rewrite it (e.g. [`pollTransaction`](https://stellar.github.io/js-stellar-sdk/module-rpc.Server.html#pollTransaction) in JS). More thinking about small "linear" apps (for lack of a better word) rather than txsub at scale.

Reply:
I guess just because we don't provide this function doesn't mean people aren't going to do this anyway. So we may as well provide it.

Reply:
> More thinking about small "linear" apps

That's a good point, and there are plenty of non-mega-scale-production use cases too, such as scripting, prototypes, etc.